### PR TITLE
feat(orc8r): Enable / Disable 5G Standalone for Pipelined and SessionD

### PR DESCRIPTION
**Pull request by pmishra-wavelabs**
## Summary

Changes for Enable / Disable 5G SA. Changes on Orc8r side.
Description : Key added "converged_access" for Sessiond and "enable_5G_feature_set" for Pipelined, and modified "5G_feature_set" with "enable_5G_feature_set".
Request Type : GET/PUT
API url : "/lte/{network_id}/cellular/epc"

## Test Plan

Update in config Files. (/var/opt/magma/configs/gateway.mconfig) 

## Additional Information

Signed-off-by: Prabhat Mishra <prabhat.mishra@wavelabs.ai>

### CODE REVIEW

File: lte/protos/mconfig/mconfigs.proto [hunk 1]
```diff
@@ -170,7 +170,10 @@ message PipelineD {
         string hmac_key = 7;
     }
     HEConfig he_config = 17;
-
+    message Enable5GFeatureSet{
+        bool enable = 1;
+    }
```
Comment (themarwhal) [inline lte/protos/mconfig/mconfigs.proto]: Do we intend to add more fields to this message in the future?

Comment (pmishra-wavelabs) [reply lte/protos/mconfig/mconfigs.proto]: Yes @themarwhal , "node_identifier" can be added in future, as I can see in pipelined.yml file.
image

File: lte/cloud/go/services/lte/obsidian/models/swagger.v1.yml [hunk 2]
```diff
@@ -1759,6 +1759,18 @@ definitions:
...
+        properties:
+          enable:
+            type: boolean
+            example: true
+            default: false
+      converged_access:
+        description: Enables 5G Standalone (SA) for sessiond at a network level
+        type: boolean
+        example: true
+        default: false
```
Comment (ulaskozat) [inline lte/cloud/go/services/lte/obsidian/models/swagger.v1.yml]: you do not need to add two more fields on swagger API to enable 5G feature. `enable_converged_core` is sufficient to figure out that you need to enable 5G services across all AGW services.

**Review by ulaskozat — CHANGES_REQUESTED** [review]
One field in swagger API will be used to enable all the services. Setting a flag in swagger API for each service separately has no utility. Either all should be enabled or none should be enabled at the same time.

File: lte/cloud/go/services/lte/obsidian/models/swagger.v1.yml (hunk 2 above)
Comment (pmishra-wavelabs) [reply lte/cloud/go/services/lte/obsidian/models/swagger.v1.yml]: Sure, @ulaskozat . Will do that change.

Comment (ulaskozat) [reply lte/cloud/go/services/lte/obsidian/models/swagger.v1.yml]: I also suggest to change the name `enable_converged_core` to `enable_5G_features`.

**Comment by uri200** [comment]
Is this the logic of the feature this setting enables already created?

**Comment by pmishra-wavelabs** [comment]
> Is this the logic of the feature this setting enables already created?

Yes @uri200 , for mme it implemented already #7729. For Pipelined and Sessiond flags needs to set.

File: lte/cloud/go/services/lte/obsidian/models/swagger.v1.yml (hunk 2 above)
Comment (pmishra-wavelabs) [reply lte/cloud/go/services/lte/obsidian/models/swagger.v1.yml]: Sure Ulas, Will do that change.

File: lte/protos/mconfig/mconfigs.proto [hunk 3]
```diff
@@ -188,6 +191,7 @@ message SessionD {
     bool gx_gy_relay_enabled = 4;
     // sentry_config stores the Sentry.io configuration for this service
     SentryConfig sentry_config = 5;
+    bool converged_access = 6;
```
Comment (hcgatewood) [inline lte/protos/mconfig/mconfigs.proto]: Enable 5g vs converged access?

File: xwf/gateway/configs/pipelined.yml [hunk 4]
```diff
@@ -144,7 +144,7 @@ has_quota_port: 51115
 no_quota_port: 51125
 
 # For 5G Functionality Support flag
-5G_feature_set:
+enable_5G_feature_set:
```
Comment (hcgatewood) [inline xwf/gateway/configs/pipelined.yml]: Here and throughout: use all-lowercase

**Review by hcgatewood — CHANGES_REQUESTED** [review]
Where are we consuming these values? I'm confused why we're setting service-level configs and mconfigs

File: cwf/gateway/configs/pipelined.yml [hunk 5]
```diff
@@ -157,7 +157,7 @@ has_quota_port: 51115
 no_quota_port: 51125
 
 # For 5G Functionality Support flag
-5G_feature_set:
+enable_5G_feature_set:
```
Comment (uri200) [inline cwf/gateway/configs/pipelined.yml]: why do we need the word `enable` here if we already have a flag called enable?

Isn't it better to leave it as 5g feature set?

Comment (uri200) [reply cwf/gateway/configs/pipelined.yml]: in case you can't use a number at the very begging, just put 5g at the end
```feature_set_5g```

**Comment by uri200** [comment]
> > Is this the logic of the feature this setting enables already created?
> 
> Yes @uri200 , for mme it implemented already #7729. For Pipelined and Sessiond flags needs to set. 

The question is: do those features already exist and work? 

We should only create enable/disable configurations if feature is there. 

If the code is there, tested and ready to be released then it is fine.

File: lte/protos/mconfig/mconfigs.proto [hunk 6]
```diff
@@ -170,7 +170,10 @@ message PipelineD {
         string hmac_key = 7;
     }
     HEConfig he_config = 17;
-
+    message FeatureSet5G{
+        bool enable = 1;
+    }
+    FeatureSet5G feature_set_5g = 18;
```
Comment (uri200) [inline lte/protos/mconfig/mconfigs.proto]: `feature_set5g` so it gets converted into `FeatureSet5g

File: lte/protos/mconfig/mconfigs.proto [hunk 7]
```diff
@@ -355,7 +359,7 @@ message MME {
     // sentry_config stores the Sentry.io configuration for this service
     SentryConfig sentry_config = 45;
     
-    bool enable_converged_core = 46;
+    bool enable_5g_features = 46;
```
Comment (uri200) [inline lte/protos/mconfig/mconfigs.proto]: `enable5g_features

File: lte/protos/mconfig/mconfigs.proto [hunk 8]
```diff
@@ -188,6 +191,7 @@ message SessionD {
     bool gx_gy_relay_enabled = 4;
     // sentry_config stores the Sentry.io configuration for this service
     SentryConfig sentry_config = 5;
+    bool enable_5g_features = 6;
```
Comment (uri200) [inline lte/protos/mconfig/mconfigs.proto]: enable5g_features

File: lte/protos/mconfig/mconfigs.proto [hunk 9]
```diff
@@ -170,7 +170,10 @@ message PipelineD {
         string hmac_key = 7;
     }
     HEConfig he_config = 17;
-
+    message FeatureSet5G{
+        bool enable = 1;
```
Comment (uri200) [inline lte/protos/mconfig/mconfigs.proto]: why do we need to wrap this inside another message

File: lte/gateway/configs/pipelined.yml [hunk 10]
```diff
@@ -188,7 +188,7 @@ ovs_internal_conntrack_port_number: 15579
 ovs_internal_conntrack_fwd_tbl_number: 202
 
 # For 5G Functionality Support flag
-5G_feature_set:
+feature_set_5g:
  enable: False
  node_identifier: 192.168.200.1
```
Comment (uri200) [inline lte/gateway/configs/pipelined.yml:193]: if this is a feature, why is not included on swagger?

File: lte/cloud/go/services/lte/servicers/builder_servicer_test.go [hunk 11]
```diff
@@ -163,6 +163,7 @@ func TestBuilder_Build(t *testing.T) {
 			SgiManagementIfaceVlan: "",
 			HeConfig:               &lte_mconfig.PipelineD_HEConfig{},
 			LiUes:                  &lte_mconfig.PipelineD_LiUes{},
+			FeatureSet_5G:          &lte_mconfig.PipelineD_FeatureSet5G{},
```
Comment (uri200) [inline lte/cloud/go/services/lte/servicers/builder_servicer_test.go]: please get used to at least create or modify one test case to verify your code

In one of those test (probably the most general), set feature to true and assert the expected result

**Review by uri200 — CHANGES_REQUESTED** [review]

**Comment by uri200** [comment]
@pshelar @koolzz please take a look at pipelined changes

File: lte/protos/mconfig/mconfigs.proto (hunk 6 above)
Comment (pmishra-wavelabs) [reply lte/protos/mconfig/mconfigs.proto]: Sure Will update.
feature_set_5g -> feature_set5g
And 
enable_5g_features -> enable5g_features
So that converted code omits underscore

File: lte/protos/mconfig/mconfigs.proto (hunk 9 above)
Comment (pmishra-wavelabs) [reply lte/protos/mconfig/mconfigs.proto]: Planned to implement "node_identifier" in future

File: lte/gateway/configs/pipelined.yml (hunk 10 above)
Comment (pmishra-wavelabs) [reply lte/gateway/configs/pipelined.yml:193]: "node_identifier" will be implemented later.

File: lte/gateway/configs/pipelined.yml [hunk 12]
```diff
@@ -188,7 +188,7 @@ ovs_internal_conntrack_port_number: 15579
 ovs_internal_conntrack_fwd_tbl_number: 202
 
 # For 5G Functionality Support flag
-5G_feature_set:
+feature_set_5g:
  enable: False
```
Comment (pshelar) [inline lte/gateway/configs/pipelined.yml:192]: the pipelineD yml file config takes precedence, so if you set any config here, pipelineD ignores mconfig. can you keep same precedence for this new feature flag.

File: lte/protos/mconfig/mconfigs.proto (hunk 6 above)
Comment (uri200) [reply lte/protos/mconfig/mconfigs.proto]: Correct. In go we use camelcase. So lets try to keep it like that if we can

File: lte/protos/mconfig/mconfigs.proto (hunk 9 above)
Comment (uri200) [reply lte/protos/mconfig/mconfigs.proto]: why is this not implemented now? I see that on pipelined.yml

Comment (pmishra-wavelabs) [reply lte/protos/mconfig/mconfigs.proto]: ok, will Implement it too.

File: lte/cloud/go/services/lte/obsidian/models/defaults.go [hunk 13]
```diff
@@ -38,7 +38,8 @@ func NewDefaultTDDNetworkConfig() *NetworkCellularConfigs {
 			GxGyRelayEnabled:         swag.Bool(false),
 			CloudSubscriberdbEnabled: false,
 			CongestionControlEnabled: swag.Bool(true),
-			EnableConvergedCore:      swag.Bool(false),
+			Enable5gFeatures:         swag.Bool(false),
+			NodeIdentifier:           "192.168.200.1",
```
Comment (uri200) [inline lte/cloud/go/services/lte/obsidian/models/defaults.go]: why don't you leave it as blank string `""`

File: lte/cloud/go/services/lte/servicers/builder_servicer_test.go [hunk 14]
```diff
@@ -1421,6 +1465,10 @@ func TestBuilder_Build_MMEPool(t *testing.T) {
 			},
 			HeConfig: &lte_mconfig.PipelineD_HEConfig{},
 			LiUes:    &lte_mconfig.PipelineD_LiUes{},
+			FeatureSet5G: &lte_mconfig.PipelineD_FeatureSet5G{
+				Enable:         false,
+				NodeIdentifier: "192.168.200.1",
+			},
```
Comment (uri200) [inline lte/cloud/go/services/lte/servicers/builder_servicer_test.go]: please go over the test and see that you do need to add an assert to check the result is what you expected. Otherwise this is not a test.

Please use [intellij](https://www.jetbrains.com/idea/) as a IDE to run the test locally, you will see how valuable those test are and how powerful the debugger it is.

**Review by uri200 — CHANGES_REQUESTED** [review]

File: lte/cloud/go/services/lte/obsidian/models/defaults.go (hunk 13 above)
Comment (pmishra-wavelabs) [reply lte/cloud/go/services/lte/obsidian/models/defaults.go]: Yes, we can put blank here. But, It requires IP address. So, I put it here. I have tested with blank entry also. If its fine with empty string I will update that.

File: lte/cloud/go/services/lte/servicers/builder_servicer_test.go (hunk 14 above)
Comment (pmishra-wavelabs) [reply lte/cloud/go/services/lte/servicers/builder_servicer_test.go]: Sure, thanks for suggestion.

**Review by pshelar — CHANGES_REQUESTED** [review]

File: lte/gateway/configs/pipelined.yml (hunk 12 above)
Comment (pmishra-wavelabs) [reply lte/gateway/configs/pipelined.yml:192]: One suggestion on that. Can I remove it from yml file. As we are assigning same as default value.

Comment (pshelar) [reply lte/gateway/configs/pipelined.yml:192]: it is nice to have local override, most of major feature have a flag in yml file.

Comment (pmishra-wavelabs) [reply lte/gateway/configs/pipelined.yml:192]: Thanks, Understood. Will update on that.

**Comment by ulaskozat** [comment]
Where are we with this PR?

**Comment by pmishra-wavelabs** [comment]
> Where are we with this PR?

Will update on this today.

File: lte/gateway/release/build-magma.sh [hunk 15]
```diff
@@ -24,7 +24,7 @@ VERSION=1.7.0 # magma version number
 SCTPD_MIN_VERSION=1.7.0 # earliest version of sctpd with which this version is compatible
 
 # RelWithDebInfo or Debug
-BUILD_TYPE=Debug
+BUILD_TYPE=RelWithDebInfo
```
Comment (ulaskozat) [inline lte/gateway/release/build-magma.sh]: why is this file modified? Looks like you need rebasing?

File: lte/gateway/release/build-magma.sh [hunk 16]
```diff
@@ -90,7 +90,7 @@ case $OS in
     ;;
     *)
     echo "Error: unknown OS option:" $OS
-    echo "Usage: [--os debian|ubuntu]"
+    echo "Usage: [--os ubuntu]"
```
Comment (ulaskozat) [inline lte/gateway/release/build-magma.sh]: same comment as above.

**Comment by ulaskozat** [comment]
@pmishra-wavelabs please do not forget to set the product label for 5g sa when you raise PRs. I added the label for this one.

File: lte/gateway/release/build-magma.sh (hunk 15 above)
Comment (pmishra-wavelabs) [reply lte/gateway/release/build-magma.sh]: > @pmishra-wavelabs please do not forget to set the product label for 5g sa when you raise PRs. I added the label for this one.

Sure

**Comment by ulaskozat** [comment]
@pmishra-wavelabs could you also post proof points that 5G services are actually running or not running after you make the config changes? E.g., enable the flag, make sure that configs are streamed down, check AGW to verify that 5G features are up or down as expected.

**Comment by ulaskozat** [comment]
@pmishra-wavelabs I do not think you did a proper rebase. A lot of files are added to the PR.

**Comment by pmishra-wavelabs** [comment]
> @pmishra-wavelabs could you also post proof points that 5G services are actually running or not running after you make the config changes? E.g., enable the flag, make sure that configs are streamed down, check AGW to verify that 5G features are up or down as expected.

Sure Will attach that.

**Comment by pmishra-wavelabs** [comment]
> @pmishra-wavelabs I do not think you did a proper rebase. A lot of files are added to the PR.

image

**Comment by pmishra-wavelabs** [comment]
> @pmishra-wavelabs I do not think you did a proper rebase. A lot of files are added to the PR.

Files under NMS. I didn't changes but it shows changes When I Pushed it. My changes are in around 20 files only.

**Comment by ulaskozat** [comment]
> > @pmishra-wavelabs I do not think you did a proper rebase. A lot of files are added to the PR.
> 
> Files under NMS. I didn't changes but it shows changes When I Pushed it. My changes are in around 20 files only.

did you update master and then rebase?

**Comment by pmishra-wavelabs** [comment]
> > > @pmishra-wavelabs I do not think you did a proper rebase. A lot of files are added to the PR.
> > 
> > 
> > Files under NMS. I didn't changes but it shows changes When I Pushed it. My changes are in around 20 files only.
> 
> did you update master and then rebase?

Yes

**Comment by pmishra-wavelabs** [comment]
> > > @pmishra-wavelabs I do not think you did a proper rebase. A lot of files are added to the PR.
> > 
> > 
> > Files under NMS. I didn't changes but it shows changes When I Pushed it. My changes are in around 20 files only.
> 
> did you update master and then rebase?

Needs to verify it again from my end. But that issue get fixed.

**Comment by pmishra-wavelabs** [comment]
> > @pmishra-wavelabs could you also post proof points that 5G services are actually running or not running after you make the config changes? E.g., enable the flag, make sure that configs are streamed down, check AGW to verify that 5G features are up or down as expected.
> 
> Sure Will attach that.

image
image
image
image

**Comment by anusha-konda** [comment]
Implemetation for seesiond and pipelined, to read 5G feature set flag from gateway.mconfig is ongoing. @prabinakpattnaik is working on it.

**Review by hcgatewood — COMMENTED** [review]
Clearing my review queue. Please re-request review when this is ready for subsequent review
